persistent/Database/Persist/Sql/TH.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskellQuotes #-}
+
+module Database.Persist.Sql.TH
+  ( fromN
+  , generateRawSqlHelper
+  , generateRawSqlInstances
+  , toN
+  )
+where
+
+import Control.Monad
+import Language.Haskell.TH
+
+#if MIN_VERSION_template_haskell(2,16,0)
+prepTupE :: a -> Maybe a
+prepTupE = Just
+#else
+prepTupE :: a -> a
+prepTupE = id
+#endif
+
+minTupleSize :: Int
+minTupleSize = 3 -- instance for 2 is already written
+
+maxTupleSize :: Int
+maxTupleSize = 62 -- GHC limitation
+
+pairedList :: [a] -> [[a]]
+pairedList [] = []
+pairedList [x] = [[x]]
+pairedList (x:y:rest) = [x, y] : pairedList rest
+
+flatTupleT :: [Name] -> Type
+flatTupleT names = foldl1 AppT $ (TupleT $ length names) : map VarT names
+
+nestedTupleT :: [Name] -> Type
+nestedTupleT names =
+   foldl1 AppT $ (TupleT $ (length names + 1) `div` 2) : map tupleOrSingleT (pairedList names)
+
+tupleOrSingleT :: [Name] -> Type
+tupleOrSingleT [a, b] = AppT (AppT (TupleT 2) (VarT a)) (VarT b)
+tupleOrSingleT [a] = VarT a
+tupleOrSingleT x = error $ "tupleOrSingleT failed to convert: " <> show x
+
+fromN :: Int -> Q (Type, Exp)
+fromN n = do
+    names <- replicateM n $ newName "x"
+    let arg = [TupP . map VarP $ names]
+    pure ( AppT (AppT ArrowT $ flatTupleT names) (nestedTupleT names)
+         , LamE arg . TupE . map (prepTupE . tupleOrSingleE) $ pairedList names
+         )
+
+    where
+        tupleOrSingleE :: [Name] -> Exp
+        tupleOrSingleE [a, b] = TupE $ map (prepTupE . VarE) [a, b]
+        tupleOrSingleE [a] = VarE a
+        tupleOrSingleE x = error $ "tupleOrSingleE failed to convert: " <> show x
+
+toN :: Int -> Q (Type, Exp)
+toN n = do
+    names <- replicateM n $ newName "x"
+    let arg = [TupP . map tupleOrSingleP $ pairedList names]
+    pure ( AppT (AppT ArrowT $ nestedTupleT names) (flatTupleT names)
+         , LamE arg . TupE $ map (prepTupE . VarE) names
+         )
+
+    where
+        tupleOrSingleP :: [Name] -> Pat
+        tupleOrSingleP [a, b] = TupP $ map VarP [a, b]
+        tupleOrSingleP [a] = VarP a
+        tupleOrSingleP x = error $ "tupleOrSingleP failed to convert: " <> show x
+
+rawSqlInstanceN :: Int -> Q Dec
+rawSqlInstanceN n = do
+    names <- replicateM n $ newName "x"
+
+    let constraints = map (\name -> AppT (ConT $ mkName "RawSql") (VarT name)) names
+        tuple = AppT (ConT $ mkName "RawSql") . foldl1 AppT $ (TupleT $ length names) : map VarT names
```

Comment:
`mkName` makes a *local* name, which can be shadowed in the environment and refer to local things. End users will need to ensure these names are in scope, which means importing the modules specifically. This isn't a great UX - generally speaking, folks don't wantt o see "Name not in scope" errors for things they didn't write!

For constant terms like this, prefer the Template Haskell quotes. 

```suggestion
        tuple = AppT (ConT $ 'RawSql) . foldl1 AppT $ (TupleT $ length names) : map VarT names
```